File: data-structures/linear/linked-list/singly-linked-list/SinglyLinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def insert_first(self, value):
        new_node = Node(value)
        new_node.next = self.head
        self.head = new_node

    def insert_last(self, value):
        new_node = Node(value)
        if not self.head:
            self.head = new_node
            return
        curr = self.head
        while curr.next:
            curr = curr.next
        curr.next = new_node

    def delete_first(self):
        if self.head:
            self.head = self.head.next

    def delete_last(self):
        if not self.head:
            return
        if not self.head.next:
            self.head = None
            return
        curr = self.head
        while curr.next and curr.next.next:
            curr = curr.next
        curr.next = None

    def insert_after(self, target, value):
        curr = self.head
        while curr:
            if curr.value == target:
                new_node = Node(value)
                new_node.next = curr.next
                curr.next = new_node
                return
            curr = curr.next
        raise ValueError(f"Value {target} not found in list.")

    def is_empty(self):
        return self.head is None

    def printList(self):
        curr = self.head
        while curr:
            print(curr.val, end="->")
            curr = curr.next
        print("None")

    def size(self):
        count = 0
        curr = self.head
        while curr:
            count += 1
            curr = curr.next
        return count

File: data-structures/linear/linked-list/singly-linked-list/SinglyLinkedList.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type LinkedList struct {
	head *Node
}

func (ll *LinkedList) AddFront(data int) {
	newNode := &Node{data: data, next: ll.head}

	if(ll.head == nil){
		ll.head = newNode
		return
	}
	newNode.next = ll.head
	ll.head = newNode
}

func (ll *LinkedList) AddBack(data int) {
	newNode := &Node{data: data}

	if ll.head == nil {
		ll.head = newNode
		return
	}

	curr := ll.head
	for curr.next != nil {
		curr = curr.next
	}
	curr.next = newNode
}

func (ll *LinkedList) DeleteFront() {
	if ll.head == nil {
		fmt.Println("List is empty.")
		return
	}
	ll.head = ll.head.next
}

func (ll *LinkedList) DeleteBack() {
	if ll.head == nil {
		fmt.Println("List is empty.")
		return
	}
	if ll.head.next == nil {
		ll.head = nil
		return
	}

	curr := ll.head
	for curr.next.next != nil {
		curr = curr.next
	}
	curr.next = nil
}

func (ll *LinkedList) Search(value int) bool {
	curr := ll.head
	for curr != nil {
		if curr.data == value {
			return true
		}
		curr = curr.next
	}
	return false
}

func (ll *LinkedList) Print() {
	curr := ll.head
	if curr == nil {
		fmt.Println("List is empty.")
		return
	}
	for curr != nil {
		fmt.Print(curr.data, " -> ")
		curr = curr.next
	}
	fmt.Println("nil")
}

func (ll *LinkedList) InsertAt(position int, data int) {
	newNode := &Node{data: data}

	if position == 0 {
		newNode.next = ll.head
		ll.head = newNode
		return
	}

	curr := ll.head
	for i := 0; i < position-1 && curr != nil; i++ {
		curr = curr.next
	}

	if curr == nil {
		fmt.Println("Position out of bounds")
		return
	}

	newNode.next = curr.next
	curr.next = newNode
}
